e2e/machine_migration_test.go

```diff
@@ -347,3 +580,76 @@ func cleanupMachineResources(ctx context.Context, cl client.Client, capiMachines
 		mapiframework.WaitForMachinesDeleted(cl, m)
 	}
 }
+
+func verifyCAPIMachineExists(cl client.Client, machineName string) {
+	Eventually(func() error {
+		_, err := capiframework.GetMachine(cl, machineName, capiframework.CAPINamespace)
+		return err
+	}, capiframework.WaitLong, capiframework.RetryLong).Should(Succeed(), "CAPI Machine should exist")
+}
+
+func verifyFieldModificationPrevented(cl client.Client, machineName string, fieldName string, modifyFunc func(*machinev1beta1.Machine)) {
+	By(fmt.Sprintf("Attempting to modify %s in MAPI Machine", fieldName))
+
+	currentMachine, err := mapiframework.GetMachine(cl, machineName)
+	if err != nil {
+		Fail(fmt.Sprintf("Failed to get current machine: %v", err))
+	}
+	updatedMachine := currentMachine.DeepCopy()
+	modifyFunc(updatedMachine)
+
+	err = cl.Update(ctx, updatedMachine)
+	Expect(err).To(HaveOccurred())
+	Expect(err.Error()).To(ContainSubstring("ValidatingAdmissionPolicy 'machine-api-machine-vap' with binding 'machine-api-machine-vap' denied request"))
+}
+
+func verifyAnnotationModificationPrevented(cl client.Client, machineName string) {
+	By("Attempting to add machine.openshift.io annotations to MAPI Machine")
+
+	currentMachine, err := mapiframework.GetMachine(cl, machineName)
+	if err != nil {
+		Fail(fmt.Sprintf("Failed to get current machine: %v", err))
+	}
+	updatedMachine := currentMachine.DeepCopy()
+	if updatedMachine.Annotations == nil {
+		updatedMachine.Annotations = make(map[string]string)
+	}
+	updatedMachine.Annotations[testAnnotationKey] = testAnnotationValue
+
+	err = cl.Update(ctx, updatedMachine)
+	Expect(err).To(HaveOccurred())
+	Expect(err.Error()).To(ContainSubstring("ValidatingAdmissionPolicy 'machine-api-machine-vap' with binding 'machine-api-machine-vap' denied request"))
+	Expect(err.Error()).To(ContainSubstring("Cannot add, modify or delete any machine.openshift.io/* annotation"))
+}
+
+// getAWSProviderSpecFromMachine extracts and unmarshals the AWS provider spec from a machine
+func getAWSProviderSpecFromMachine(machine *machinev1beta1.Machine) (*machinev1beta1.AWSMachineProviderConfig, error) {
+	if machine.Spec.ProviderSpec.Value == nil {
```

Comment:
We could wrap https://github.com/openshift/cluster-capi-operator/blob/e190ae0ad574ba0017f94800b259fb797d88813e/pkg/conversion/mapi2capi/aws.go#L347-L359, and then we're not duplicating code :)